src/libraries/Common/tests/System/IO/ConnectedStreams.cs

```diff
@@ -0,0 +1,169 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO
+{
+    /// <summary>Provides support for in-memory producer/consumer streams.</summary>
+    internal sealed class ConnectedStreams
+    {
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional() =>
+            CreateBidirectional(StreamBuffer.DefaultInitialBufferSize, StreamBuffer.DefaultMaxBufferSize);
+
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <param name="initialBufferSize">The initial buffer size to use when storing data in the connection.</param>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional(int initialBufferSize) =>
+            CreateBidirectional(initialBufferSize, StreamBuffer.DefaultMaxBufferSize);
+
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <param name="initialBufferSize">The initial buffer size to use when storing data in the connection.</param>
+        /// <param name="maxBufferSize">
+        /// The maximum buffer size to use when storing data in the connection.  When this limit is reached,
+        /// writes will block until additional space becomes available.
+        /// </param>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional(int initialBufferSize, int maxBufferSize)
+        {
+            // Each direction needs a buffer; one stream will use b1 for reading and b2 for writing,
+            // and the other stream will do the inverse.
+            var b1 = new StreamBuffer(initialBufferSize, maxBufferSize);
+            var b2 = new StreamBuffer(initialBufferSize, maxBufferSize);
+
+            // Both StreamBuffers are shared between the streams: we don't want to dispose of the underlying storage
+            // in the stream buffer until both streams are done with them, so we ref count and only dispose
+            // when both streams have been disposed.  To share the same integer, it's put onto the heap.
+            var refCount = new StrongBox<int>(2);
```

Comment:
This is a bit of a nit, but I wonder if it's clearer and simpler to just have an internal class like StreamBufferState or similar that holds the two StreamBuffers and the refCount. You're allocating the refcount on the heap anyway.

You would need to distinguish in the streams themselves which side is which, with a `bool _isClient` or something like that, and you would need to test this on each read/write etc to determine which StreamBuffer to access. So there's a tradeoff here.

I'm fine with either approach, just thought I'd mention it.

Reply:
Thanks.  To me this is cleaner.  But if there ends up being more state in the future that's necessary, I can see such an approach being useful.